### DISCUSSION TITLE
How the registered services in Setup module are OFF after the installation

### DISCUSSION THREAD

**Discussion by hishamco** [OrchardCMS/OrchardCore#13729] [Q&A] [answered]
@jtkech could you please tell me how the registered services are in `OC.Setup` module are removed from DI after the installation process is done. Say for instance `ILocalizationFileLocationProvider`

**Comment by ns8482e** [comment]
You mean after setup? 

Site is configure based on setup recipe. Services are not available if module is not enabled by setup recipe.

If localization feature is not enabled by recipe then related services will not be available in the site.

**Reply by jtkech** [reply]
When a tenant is `Uninitialized` we create a setup context, for example here.

https://github.com/OrchardCMS/OrchardCore/blob/edccca0b92228d55ec70ef32deda5c47e5bdfb99/src/OrchardCore/OrchardCore/Shell/ShellHost.cs#L369

This context is built with a `MinimumShellDescriptor()` which is only composed of the `ShellFeature`s (no recipe) that was registered at the host level including `OC.Setup` and that below we resolve through `_shellFeatures`.

https://github.com/OrchardCMS/OrchardCore/blob/edccca0b92228d55ec70ef32deda5c47e5bdfb99/src/OrchardCore/OrchardCore/Shell/Builders/ShellContextFactory.cs#L99-L108

To register the setup feature we use `.AddSetupFeatures("OrchardCore.Setup")` which is called in `AddOrchardCms()`.

https://github.com/OrchardCMS/OrchardCore/blob/edccca0b92228d55ec70ef32deda5c47e5bdfb99/src/OrchardCore/OrchardCore/Modules/Extensions/OrchardCoreBuilderExtensions.cs#L47-L55

So, if the tenant is `Uninitialized` the shell descriptor will have the setup feature whose startup will be executed, then when the tenant is setup a new shell descriptor is built including the features of the setup recipe.

As @ns8482e said here it depends on the setup recipe that you have used, for example if `OC.Setup` is part of the recipe features, its startup will be still executed, otherwise not.

When a tenant is running (has been setup) we still use the `ShellFeature`s registered in the DI but we only use those that are marked `AlwaysEnabled`, which is not the case of `OC.Setup` when it is added by `.AddSetupFeatures()` but which is the case when features are added for example by `AddGlobalFeatures()`.

**Comment by hishamco** [comment]
> You mean after setup?

After the setup process is completed

> Site is configure based on setup recipe. Services are not available if module is not enabled by setup recipe.

> If localization feature is not enabled by recipe then related services will not be available in the site.

I knew I was not talking about recipes here. Let me clarify `OC.Setup` allow users to choose the language - even the localization module is not enabled - that's because there's a services light up to enable the localization, my question is how the `ILocalizationFileLocationProvider` service for instance is not in DI after the tenant is ready?

@jtketch thanks for the clarification, but seems my question wasn't clear before, please refer to the above paragraph

**Comment by hishamco** [comment]
> You mean after setup?

After the setup process is completed

> Site is configure based on setup recipe. Services are not available if module is not enabled by setup recipe.

> If localization feature is not enabled by recipe then related services will not be available in the site.

I knew I was not talking about recipes here. Let me clarify `OC.Setup` allow users to choose the language - even the localization module is not enabled - that's because there's a services light up to enable the localization, my question is how the `ILocalizationFileLocationProvider` service for instance is not in DI after the tenant is ready?

@jtketch thanks for the clarification, but seems my question wasn't clear before, please refer to the above paragraph

**Reply by jtkech** [reply, marked as answer]
As I understand the question both @ns8482e and me already answered, after the setup the Tenant is rebuilt but without the OC.Setup feature whose startup is not executed, so the related services are no longer part of the Tenant container.

**Reply by hishamco** [reply]
> after the setup the Tenant is rebuilt but without the OC.Setup feature whose startup is not executed, so the related services are no longer part of the Tenant container.

Thanks, JT maybe the answer wasn't clear to me at the beginning :) could you please point me where the tenant is rebuilt

**Comment by jtkech** [comment]
At the end of `SetupInternalAsync()` => `_shellHost.UpdateShellSettingsAsync()` => `ReloadShellContextAsync()` => which replaces the current context in `_shellContexts` by a new `ShellContext.PlaceHolder { Settings = settings }`.

A `PlaceHolder` is not a fully built shell context but it is marked as `Released`, so in a subsequent request/redirection it will be built by using the shell settings that are up to date (e.g. tenant is running) and by using the shell descriptor retrieved from the database, so not a `MinimumShellDescriptor` but the one that was updated during the setup.

**Reply by hishamco** [reply]
Thanks a lot
